src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs

```diff
@@ -438,29 +438,126 @@ public void OnWindowUpdate(int amount)
+                (KnownHeaders.Expect.Descriptor, Array.Empty<byte>()),
+                (KnownHeaders.Expires.Descriptor, Array.Empty<byte>()),
+                (KnownHeaders.From.Descriptor, Array.Empty<byte>()),
+                (KnownHeaders.Host.Descriptor, Array.Empty<byte>()),
+                (KnownHeaders.IfMatch.Descriptor, Array.Empty<byte>()),
+                (KnownHeaders.IfModifiedSince.Descriptor, Array.Empty<byte>()),
+                (KnownHeaders.IfNoneMatch.Descriptor, Array.Empty<byte>()),
+                (KnownHeaders.IfRange.Descriptor, Array.Empty<byte>()),
+                (KnownHeaders.IfUnmodifiedSince.Descriptor, Array.Empty<byte>()),
+                (KnownHeaders.LastModified.Descriptor, Array.Empty<byte>()),
+                (KnownHeaders.Link.Descriptor, Array.Empty<byte>()),
+                (KnownHeaders.Location.Descriptor, Array.Empty<byte>()),
+                (KnownHeaders.MaxForwards.Descriptor, Array.Empty<byte>()),
+                (KnownHeaders.ProxyAuthenticate.Descriptor, Array.Empty<byte>()),
+                (KnownHeaders.ProxyAuthorization.Descriptor, Array.Empty<byte>()),
+                (KnownHeaders.Range.Descriptor, Array.Empty<byte>()),
+                (KnownHeaders.Referer.Descriptor, Array.Empty<byte>()),
+                (KnownHeaders.Refresh.Descriptor, Array.Empty<byte>()),
+                (KnownHeaders.RetryAfter.Descriptor, Array.Empty<byte>()),
+                (KnownHeaders.Server.Descriptor, Array.Empty<byte>()),
+                (KnownHeaders.SetCookie.Descriptor, Array.Empty<byte>()),
+                (KnownHeaders.StrictTransportSecurity.Descriptor, Array.Empty<byte>()),
+                (KnownHeaders.TransferEncoding.Descriptor, Array.Empty<byte>()),
+                (KnownHeaders.UserAgent.Descriptor, Array.Empty<byte>()),
+                (KnownHeaders.Vary.Descriptor, Array.Empty<byte>()),
+                (KnownHeaders.Via.Descriptor, Array.Empty<byte>()),
+                (KnownHeaders.WWWAuthenticate.Descriptor, Array.Empty<byte>()),
+            };
+
             void IHttpHeadersHandler.OnStaticIndexedHeader(int index)
             {
-                // TODO: https://github.com/dotnet/runtime/issues/1505
-                ref readonly HeaderField entry = ref H2StaticTable.Get(index - 1);
-                OnHeader(entry.Name, entry.Value);
+                Debug.Assert(index >= FirstHPackRequestPseudoHeaderId && index <= LastHPackNormalHeaderId);
+
+                if (index <= LastHPackRequestPseudoHeaderId)
+                {
+                    if (NetEventSource.Log.IsEnabled()) Trace($"Invalid request pseudo-header ID {index}.");
+                    throw new HttpRequestException(SR.net_http_invalid_response);
+                }
+                else if (index <= LastHPackStatusPseudoHeaderId)
+                {
+                    int statusCode = HPackStaticStatusCodeTable[index - FirstHPackStatusPseudoHeaderId];
```

Comment:
nit: Replace the array lookup with a switch?

Reply:
That might be slightly better but I'm not sure it matters in practice.

Reply:
A switch does seem a little better here to me as well, but not a big deal.